gitlab/resource_gitlab_managed_license.go

```diff
@@ -0,0 +1,168 @@
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				// GitLab's edit API doesn't allow us to edit the name, only
+				// the approval status.
+				ForceNew:    true,
+				Description: "The name of the managed license (I.e., 'Apache License 2.0' or 'MIT license')",
+			},
+			"approval_status": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     false,
+				ValidateFunc: validation.StringInSlice([]string{"approved", "blacklisted"}, true),
+				Description:  "Whether the license is approved or not. Only 'approved' or 'blacklisted' allowed.",
+			},
+		},
+	}
+}
+
+func resourceGitlabManagedLicenseImporter(ctx context.Context, data *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	parts := strings.SplitN(data.Id(), ":", 2)
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid label id (should be <project id>:<license id>). Input: %s", data.Id())
+	}
+
+	data.SetId(parts[1])
+	data.Set("project", parts[0])
+
+	diagnostic := resourceGitlabManagedLicenseRead(ctx, data, meta)
+	if diagnostic.HasError() {
+		return nil, fmt.Errorf("failed to managed license instance %s: %s", data.Id(), diagnostic[0].Summary)
+	}
+
+	return []*schema.ResourceData{data}, nil
+}
+
+func resourceGitlabManagedLicenseCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+
+	options := &gitlab.AddManagedLicenseOptions{
+		Name:           gitlab.String(d.Get("name").(string)),
+		ApprovalStatus: stringToApprovalStatus(d.Get("approval_status").(string)),
+	}
+
+	log.Printf("[DEBUG] create gitlab Managed License on Project %s, with Name %s", project, *options.Name)
+
+	addManagedLicense, _, err := client.ManagedLicenses.AddManagedLicense(project, options, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(strconv.Itoa(addManagedLicense.ID))
+	return resourceGitlabManagedLicenseRead(ctx, d, meta)
+}
+
+func resourceGitlabManagedLicenseDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	licenseId, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("%s cannot be converted to int", d.Id()))
+	}
+
+	log.Printf("[DEBUG] Delete gitlab Managed License %s", d.Id())
+	_, err = client.ManagedLicenses.DeleteManagedLicense(project, licenseId, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("failed to delete managed license %d for projec %s: %w", licenseId, project, err))
+	}
+
+	return nil
+}
+
+func resourceGitlabManagedLicenseRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	id, err := strconv.Atoi(d.Id())
```

Comment:
Minor request, to keep the naming style used in the other methods, would you mind renaming this to `licenseId, err := strconv.Atoi(d.Id())` please?

Reply:
@PatrickRice-KSC Most of our project resources follow the ID convention of `<project-id>:<resource-specific-id>` (like you have used in  your importer).

Can you remove your custom importer and use

```go
Importer: &schema.ResourceImporter{
	StateContext: resourceGitlabDeployKeyEnableStateImporter,
},
```

And then parse the project ID and license ID from the ID in this *Read function instead of getting the project ID from `d.Get("project")`.

(There's a section in the CONTRIBUTING.md about resource ID guidance as well.)

Reply:
@armsnyder  - I don't have any problem doing that, but isn't that a bit confusing to use the import method defined in the deploy key enable resource for the managed license resource? If we did that, seems like we'd want to refactor that method into a utility class since it's not specific to the deploy key resource anymore. 

That method also sets "key_id" and "project" into the ResourceData, so I would then have to modify the `resourceGitlabManagedLicenseRead` method to check both `d.Id()` or `d.Get("key_id")`. Certainly workable, just odd given "key_id" isn't used anywhere else in managed license. Maybe I'm missing something?

@nagyv  - for sure, I'll update that while I wait from the response from armsnyder :)

Reply:
Nevermind, I'm silly and I see where it's setting the Id (not sure how I missed that, to be honest...), so I can easily update Read to parse based on the presence of a ":" in the Id. I think the rest of the question is still valid though.

Reply:
Oops I thought I was copying

```go
Importer: &schema.ResourceImporter{
    StateContext: schema.ImportStatePassthroughContext,
},
```

(from the CONTRIBUTING.md) That's what I meant.

Reply:
@armsnyder  - Gotcha, I did try that (specifically because it was mentioned in the CONTRIBUTING.md) but it throws an error that it can't translate "1234:5678" (whatever the keys are from the test obv) into an integer. Lemme work on it a bit and see if I can get it to work.

Reply:
Got it - had to update all the CRUD methods to parse out instead of using the attributes from the data object, and then it works. Everything is updated, and tests are passing again.